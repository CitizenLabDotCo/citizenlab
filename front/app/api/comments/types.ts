import { IRelationship, Multiloc, ILinks } from 'typings';
import { Keys } from 'utils/cl-react-query/types';
import commentsKeys from './keys';

export type CommentsKeys = Keys<typeof commentsKeys>;

export type ICommentParameters = {
  ideaId?: string;
  initiativeId?: string;
  userId?: string;
  commentId?: string;
};

interface CommentAttributes {
  upvotes_count: number;
  downvotes_count: number;
  created_at: string;
  updated_at: string;
  children_count: number;
}

interface IPresentComment extends CommentAttributes {
  body_multiloc: Multiloc;
  publication_status: 'published';
  anonymous?: boolean;
  author_hash?: string;
}

interface IDeletedComment extends CommentAttributes {
  body_multiloc: null;
  publication_status: 'deleted';
}

export interface ICommentData {
  id: string;
  type: 'comment';
  attributes: IPresentComment | IDeletedComment;
  relationships: {
    post: {
      data: IRelationship;
    };
    author: {
      data: IRelationship | null;
    };
    parent: {
      data: IRelationship | null;
    };
    user_vote?: {
      data: IRelationship | null;
    };
  };
}

export interface IComment {
  data: ICommentData;
}

export interface IComments {
  data: ICommentData[];
  links: ILinks;
}

export interface INewComment {
  ideaId?: string;
  initiativeId?: string;
  author_id: string;
  parent_id?: string;
  body_multiloc: Multiloc;
  anonymous?: boolean;
}

export interface IUpdatedComment {
  commentId: string;
  author_id?: string;
  parent_id?: string;
  body_multiloc: Multiloc;
}

export const DeleteReasonCode = {
  irrelevant: 'irrelevant',
  inappropriate: 'inappropriate',
  other: 'other',
};

export interface DeleteReason {
  reason_code: keyof typeof DeleteReasonCode;
  other_reason: string | null;
}

// back-end also offers 'upvotes_count', but we haven't implemented it
export type CommentsSort = '-new' | 'new' | '-upvotes_count';

export type ICommentQueryParameters = {
  sort?: CommentsSort;
  pageNumber?: number;
  pageSize?: number;
};
