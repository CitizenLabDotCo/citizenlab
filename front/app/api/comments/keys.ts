import { QueryKeys } from 'utils/cl-react-query/types';
import { ICommentParameters, ICommentQueryParameters } from './types';

const baseKey = { type: 'comment' };

const commentKeys = {
  all: () => [baseKey],
  lists: () => [{ ...baseKey, operation: 'list' }],
  list: (params: ICommentParameters & ICommentQueryParameters) => [
    { ...baseKey, operation: 'list', parameters: params },
  ],
  items: () => [{ ...baseKey, operation: 'item' }],
  item: ({ id }: { id: string }) => [
    {
      ...baseKey,
      operation: 'item',
      parameters: { id },
    },
  ],
} satisfies QueryKeys;

export default commentKeys;
