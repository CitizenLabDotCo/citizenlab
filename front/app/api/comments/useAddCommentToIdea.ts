import { useMutation, useQueryClient } from '@tanstack/react-query';
import commentKeys from 'api/comments/keys';
import ideasKeys from 'api/ideas/keys';
import { API_PATH } from 'containers/App/constants';
import { CLErrors } from 'typings';
import fetcher from 'utils/cl-react-query/fetcher';
import streams from 'utils/streams';
import { INewComment, IComment } from './types';

const addCommentToIdea = async ({ ideaId, ...requestBody }: INewComment) =>
  fetcher<IComment>({
    path: `/ideas/${ideaId}/comments`,
    action: 'post',
    body: { comment: { ...requestBody } },
  });

const useAddCommentToIdea = () => {
  const queryClient = useQueryClient();
  return useMutation<IComment, CLErrors, INewComment>({
    mutationFn: addCommentToIdea,
    onSuccess: (_data, variables) => {
      queryClient.invalidateQueries({
        queryKey: commentKeys.list({ userId: variables.author_id }),
      });
      queryClient.invalidateQueries({
        queryKey: commentKeys.list({ ideaId: variables.ideaId }),
      });
      queryClient.invalidateQueries({
        queryKey: ideasKeys.item({ id: variables.ideaId }),
      });

      streams.fetchAllWith({
        apiEndpoint: [
          `${API_PATH}/users/${variables.author_id}/comments_count`,
        ],
      });
    },
  });
};

export default useAddCommentToIdea;
