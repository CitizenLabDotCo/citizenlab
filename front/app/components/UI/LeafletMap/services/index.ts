import { init, addTileLayer, changeView } from './setup';

import { addLayersControl, removeLayersControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import {
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
} from './markers';

export default {
  init,
  addTileLayer,
  changeView,
  addLayersControl,
  removeLayersControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
};
