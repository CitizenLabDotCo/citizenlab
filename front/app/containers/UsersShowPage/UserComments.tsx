import React, { memo } from 'react';
import { isNilOrError } from 'utils/helperUtils';
import { adopt } from 'react-adopt';
import { groupBy } from 'lodash-es';

// components
import PostCommentGroup from './PostCommentGroup';
import Button from 'components/UI/Button';

// resources
import GetCommentsForUser, {
  GetCommentsForUserChildProps,
} from 'resources/GetCommentsForUser';
import GetAuthUser, { GetAuthUserChildProps } from 'resources/GetAuthUser';

// style
import styled, { withTheme } from 'styled-components';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from './messages';
import { darken, rgba } from 'polished';
import { media, colors, fontSizes } from 'utils/styleUtils';
import { ScreenReaderOnly } from 'utils/a11y';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;

  max-width: 760px;
`;

const Footer = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  margin-top: 30px;

  ${media.phone`
    flex-direction: column;
    align-items: stretch;
    margin-top: 0px;
  `}
`;

const MessageContainer = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  color: ${colors.textSecondary};
  font-size: ${fontSizes.m}px;
  font-weight: 400;
`;

const LoadMoreButton = styled(Button)``;

interface InputProps {
  userId: string;
}

interface DataProps {
  comments: GetCommentsForUserChildProps;
  authUser: GetAuthUserChildProps;
}

interface Props extends InputProps, DataProps {
  theme: any;
}

export const UserComments = memo<Props>(
  ({ comments, userId, theme, authUser }) => {
    if (!isNilOrError(comments)) {
      const { commentsList } = comments;

      if (commentsList === undefined) {
        return (
          <MessageContainer>
            <FormattedMessage {...messages.loadingComments} />
          </MessageContainer>
        );
      }

      if (
        commentsList === null ||
        (!isNilOrError(commentsList) && commentsList.length === 0)
      ) {
        if (!isNilOrError(authUser) && userId === authUser.id) {
          return (
            <MessageContainer>
              <FormattedMessage {...messages.noCommentsForYou} />
            </MessageContainer>
          );
        }

        return (
          <MessageContainer>
            <FormattedMessage {...messages.noCommentsForUser} />
          </MessageContainer>
        );
      }

      if (!isNilOrError(commentsList) && commentsList.length > 0) {
        const commentGroups = groupBy(
          commentsList,
          (comment) => comment.relationships.post.data.id
        );

        return (
          <Container className="e2e-profile-comments">
            <ScreenReaderOnly>
              <FormattedMessage
                tagName="h2"
                {...messages.invisibleTitleUserComments}
              />
            </ScreenReaderOnly>
            <>
              {Object.keys(commentGroups).map((postId) => {
                const commentGroup = commentGroups[postId];
                const postType = commentGroup[0].relationships.post.data
                  .type as 'idea' | 'initiative';

                return (
                  <PostCommentGroup
                    key={postId}
                    postId={postId}
                    postType={postType}
                    comments={commentGroup}
                    userId={userId}
                  />
                );
              })}
            </>

            {comments.hasMore && (
              <Footer>
                <LoadMoreButton
                  onClick={comments.loadMore}
                  processing={comments.loadingMore}
                  icon="refresh"
                  textColor={theme.colors.tenantText}
                  textHoverColor={darken(0.1, theme.colors.tenantText)}
                  bgColor={rgba(theme.colors.tenantText, 0.08)}
                  bgHoverColor={rgba(theme.colors.tenantText, 0.12)}
                  height="50px"
                >
                  <FormattedMessage {...messages.loadMoreComments} />
                </LoadMoreButton>
              </Footer>
            )}
          </Container>
        );
      }
    }

    return (
      <MessageContainer>
        <FormattedMessage {...messages.tryAgain} />
      </MessageContainer>
    );
  }
);

const Data = adopt<DataProps, InputProps>({
  comments: ({ userId, render }) => (
    <GetCommentsForUser userId={userId}>{render}</GetCommentsForUser>
  ),
  authUser: <GetAuthUser />,
});

const UserCommentsWithHocs = withTheme(UserComments);

const WrappedUserComments = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {(dataProps) => <UserCommentsWithHocs {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedUserComments;
