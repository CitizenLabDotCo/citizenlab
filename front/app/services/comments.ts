import { API_PATH } from 'containers/App/constants';
import request from 'utils/request';
import streams, { IStreamParams } from 'utils/streams';
import { IRelationship, Multiloc } from 'typings';

interface CommentAttributes {
  upvotes_count: number;
  downvotes_count: number;
  created_at: string;
  updated_at: string;
  children_count: number;
}

interface IPresentComment extends CommentAttributes {
  body_multiloc: Multiloc;
  publication_status: 'published';
}

interface IDeletedComment extends CommentAttributes {
  body_multiloc: null;
  publication_status: 'deleted';
}

export interface ICommentData {
  id: string;
  type: 'comment';
  attributes: IPresentComment | IDeletedComment;
  relationships: {
    post: {
      data: IRelationship;
    };
    author: {
      data: IRelationship | null;
    };
    parent: {
      data: IRelationship | null;
    };
    user_vote?: {
      data: IRelationship | null;
    };
  };
}

export interface ICommentLinks {
  self: string;
  first: string;
  prev: string;
  next: string;
  last: string;
}

export interface IComment {
  data: ICommentData;
}

export interface IComments {
  data: ICommentData[];
  links: ICommentLinks;
}

export interface IUpdatedComment {
  author_id?: string;
  parent_id?: string;
  body_multiloc: { [key: string]: string };
}

export const DeleteReasonCode = {
  irrelevant: 'irrelevant',
  inappropriate: 'inappropriate',
  other: 'other',
};

export interface DeleteReason {
  reason_code: keyof typeof DeleteReasonCode;
  other_reason: string | null;
}

export type CommentsSort = '-new' | 'upvotes_count' | 'new' | '-upvotes_count';

export function childCommentsStream(
  commentId: string,
  streamParams: IStreamParams | null = null
) {
  return streams.get<IComments>({
    apiEndpoint: `${API_PATH}/comments/${commentId}/children`,
    ...streamParams,
  });
}

export async function addCommentToIdea(
  ideaId: string,
  projectId: string,
  authorId: string,
  body: { [key: string]: string }
) {
  const comment = await streams.add<IComment>(
    `${API_PATH}/ideas/${ideaId}/comments`,
    {
      comment: {
        author_id: authorId,
        body_multiloc: body,
      },
    },
    true
  );

  // refetch commentsForUser and comments for user count
  streams.fetchAllWith({
    apiEndpoint: [
      `${API_PATH}/users/${authorId}/comments`,
      `${API_PATH}/users/${authorId}/comments_count`,
    ],
    dataId: [ideaId, projectId, comment.data.id],
  });

  return comment;
}

export async function addCommentToInitiative(
  initiativeId: string,
  authorId: string,
  body: { [key: string]: string }
) {
  const comment = await streams.add<IComment>(
    `${API_PATH}/initiatives/${initiativeId}/comments`,
    {
      comment: {
        author_id: authorId,
        body_multiloc: body,
      },
    },
    true
  );

  // refetch commentsForUser and comments for user count
  streams.fetchAllWith({
    apiEndpoint: [
      `${API_PATH}/users/${authorId}/comments`,
      `${API_PATH}/users/${authorId}/comments_count`,
    ],
    dataId: [initiativeId, comment.data.id],
  });

  return comment;
}

export async function addCommentToIdeaComment(
  ideaId: string,
  projectId: string,
  authorId: string,
  parentCommentId: string,
  body: { [key: string]: string },
  waitForChildCommentsRefetch = false
) {
  const comment = await streams.add<IComment>(
    `${API_PATH}/ideas/${ideaId}/comments`,
    {
      comment: {
        author_id: authorId,
        parent_id: parentCommentId,
        body_multiloc: body,
      },
    },
    true
  );

  // refetch commentsForUser and comments for user count
  streams.fetchAllWith({
    apiEndpoint: [
      `${API_PATH}/users/${authorId}/comments`,
      `${API_PATH}/users/${authorId}/comments_count`,
    ],
  });

  if (waitForChildCommentsRefetch) {
    await streams.fetchAllWith({
      apiEndpoint: [`${API_PATH}/comments/${parentCommentId}/children`],
      dataId: [ideaId, projectId, parentCommentId, comment.data.id],
    });
  } else {
    streams.fetchAllWith({
      dataId: [ideaId, projectId, parentCommentId, comment.data.id],
      apiEndpoint: [`${API_PATH}/comments/${parentCommentId}/children`],
    });
  }

  return comment;
}

export async function addCommentToInitiativeComment(
  initiativeId: string,
  authorId: string,
  parentCommentId: string,
  body: { [key: string]: string },
  waitForChildCommentsRefetch = false
) {
  const comment = await streams.add<IComment>(
    `${API_PATH}/initiatives/${initiativeId}/comments`,
    {
      comment: {
        author_id: authorId,
        parent_id: parentCommentId,
        body_multiloc: body,
      },
    },
    true
  );

  // refetch commentsForUser and comments for user count
  streams.fetchAllWith({
    apiEndpoint: [
      `${API_PATH}/users/${authorId}/comments`,
      `${API_PATH}/users/${authorId}/comments_count`,
    ],
  });

  if (waitForChildCommentsRefetch) {
    await streams.fetchAllWith({
      apiEndpoint: [`${API_PATH}/comments/${parentCommentId}/children`],
      dataId: [initiativeId, parentCommentId, comment.data.id],
    });
  } else {
    streams.fetchAllWith({
      dataId: [initiativeId, parentCommentId, comment.data.id],
      apiEndpoint: [`${API_PATH}/comments/${parentCommentId}/children`],
    });
  }

  return comment;
}

export async function updateComment(
  commentId: string,
  object: IUpdatedComment
) {
  const response = await streams.update<IComment>(
    `${API_PATH}/comments/${commentId}`,
    commentId,
    { comment: object }
  );

  // refetch commentsForUser
  if (object.author_id) {
    streams.fetchAllWith({
      apiEndpoint: [`${API_PATH}/users/${object.author_id}/comments`],
    });
  }

  return response;
}

export async function markForDeletion(
  commentId: string,
  authorId?: string,
  projectId?: string | null,
  reason?: DeleteReason
) {
  if (reason && reason.reason_code !== 'other') {
    reason.other_reason = null;
  }

  const response = await request(
    `${API_PATH}/comments/${commentId}/mark_as_deleted`,
    { comment: reason },
    { method: 'POST' },
    {}
  );
  const dataIdsToRefetch = projectId ? [commentId, projectId] : [commentId];
  const apiEndpointsToRefetch = authorId
    ? [
        `${API_PATH}/users/${authorId}/comments`,
        `${API_PATH}/users/${authorId}/comments_count`,
      ]
    : [];

  await streams.fetchAllWith({
    dataId: dataIdsToRefetch,
    apiEndpoint: apiEndpointsToRefetch,
  });

  return response;
}
