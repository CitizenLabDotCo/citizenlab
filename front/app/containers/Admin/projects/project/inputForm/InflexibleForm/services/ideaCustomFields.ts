import { API_PATH } from 'containers/App/constants';
import streams, { IStreamParams } from 'utils/streams';
import { Multiloc } from 'typings';
import { CustomFieldCodes } from 'services/ideaCustomFieldsSchemas';

export type Visibility = 'admins' | 'public';

export interface IIdeaCustomFieldData {
  id: string;
  type: 'custom_field';
  attributes: {
    key: string;
    input_type:
      | 'text'
      | 'number'
      | 'multiline_text'
      | 'html'
      | 'text_multiloc'
      | 'multiline_text_multiloc'
      | 'html_multiloc'
      | 'select'
      | 'multiselect'
      | 'checkbox'
      | 'date'
      | 'files'
      | 'image_files';
    title_multiloc: Multiloc;
    description_multiloc: Multiloc;
    required: boolean;
    ordering: null;
    enabled: boolean;
    code: CustomFieldCodes;
    created_at: null;
    updated_at: null;
    visible_to: Visibility;
  };
}

export interface IIdeaCustomField {
  data: IIdeaCustomFieldData;
}

export interface IIdeaCustomFields {
  data: IIdeaCustomFieldData[];
}

export interface IUpdatedIdeaCustomFieldProperties {
  description_multiloc?: Multiloc;
  enabled?: boolean;
  required?: boolean;
  visible_to?: Visibility;
}

export function ideaCustomFieldsStream(
  projectId: string,
  streamParams: IStreamParams | null = null
) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  return streams.get<IIdeaCustomFields>({ apiEndpoint, ...streamParams });
}

export async function updateIdeaCustomField(
  projectId: string,
  ideaCustomFieldId: string,
  code: string | undefined,
  object: IUpdatedIdeaCustomFieldProperties
) {
  const apiEndpoint = code
    ? `${API_PATH}/admin/projects/${projectId}/custom_fields/by_code/${code}`
    : `${API_PATH}/admin/projects/${projectId}/custom_fields/update/${ideaCustomFieldId}`;
  const updateObject = { custom_field: object };
  return streams.update<IIdeaCustomField>(
    apiEndpoint,
    ideaCustomFieldId,
    updateObject
  );
}

export async function refetchCustomFields(projectId) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  streams.fetchAllWith({ apiEndpoint: [apiEndpoint] });
}
