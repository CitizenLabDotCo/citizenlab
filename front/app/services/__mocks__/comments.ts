import { ICommentData, IComment, IComments } from 'services/comments';
import { BehaviorSubject } from 'rxjs';

export const makeCommentData = (
  id = 'commentId',
  attributes = {},
  ideaId?: string,
  authorId?: string,
  parentId?: string
): ICommentData => ({
  id,
  type: 'comment',
  attributes: {
    body_multiloc: { en: 'Just a comment' },
    publication_status: 'published',
    upvotes_count: 0,
    downvotes_count: 0,
    created_at: '2019-03-12T00: 00: 00.000Z',
    updated_at: '2019-03-26T14: 32: 32.000Z',
    children_count: 0,
    ...attributes,
  },
  relationships: {
    post: {
      data: {
        id: ideaId ? ideaId : 'ideaId',
        type: 'idea',
      },
    },
    author: {
      data: {
        id: authorId ? authorId : 'authorId',
        type: 'user',
      },
    },
    parent: {
      data: parentId
        ? {
            id: parentId,
            type: 'comment',
          }
        : null,
    },
    user_vote: {
      data: null,
    },
  },
});

export const makeComment = (
  id = 'commentId',
  attributes = {},
  ideaId?: string,
  authorId?: string,
  parentId?: string
): IComment => ({
  data: makeCommentData(id, attributes, ideaId, authorId, parentId),
});

export const makeComments = (
  argumentsArray = [
    {
      id: undefined,
      attributes: {},
      ideaId: undefined,
      authorId: undefined,
      parentId: undefined,
    },
  ]
): IComments => ({
  data: argumentsArray.map((args) =>
    makeCommentData(
      args.id,
      args.attributes,
      args.ideaId,
      args.authorId,
      args.parentId
    )
  ),
  links: {
    self: 'self',
    first: 'first',
    last: 'last',
    prev: 'prev',
    next: 'next',
  },
});

const mockCommentsForUser = new BehaviorSubject<undefined | IComments>(
  undefined
);

export const __setMockCommentsForUser = (comments: IComments) => {
  mockCommentsForUser.next(comments);
};

export const commentsForUserStream = jest.fn(() => {
  return {
    observable: mockCommentsForUser,
  };
});
