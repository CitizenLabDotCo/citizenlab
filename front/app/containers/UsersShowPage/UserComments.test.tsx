// @ts-nocheck
// libraries
import React from 'react';
import { shallow } from 'enzyme';

import { theme } from 'utils/testUtils/withTheme';

// component to test
import { UserComments } from './UserComments';

// mock utilities
jest.mock('utils/cl-intl');
jest.mock('services/comments');
jest.mock('modules', () => ({ streamsToReset: [] }));
jest.mock('components/Outlet', () => 'Outlet');

import { makeComments } from 'services/comments';

describe('<UserComments />', () => {
  it('renders correctly when empty', () => {
    const commentsAsReturned = { commentsList: [] };
    const Wrapper = shallow(<UserComments comments={commentsAsReturned} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly when empty and own profile', () => {
    const commentsAsReturned = { commentsList: [] };
    const Wrapper = shallow(
      <UserComments
        comments={commentsAsReturned}
        userId="someUser"
        authUser={{ id: 'someUser' }}
      />
    );
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly when error', () => {
    const commentsAsReturned = { commentsList: new Error() };
    const Wrapper = shallow(<UserComments comments={commentsAsReturned} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly with actual comments', () => {
    const commentsAsReturned = {
      commentsList: makeComments([
        { ideaId: 'idea1' },
        { ideaId: 'idea2' },
        { ideaId: 'idea2' },
      ]).data,
    };
    const Wrapper = shallow(<UserComments comments={commentsAsReturned} />);
    expect(Wrapper).toMatchSnapshot();
  });
  it('renders correctly with actual comments and load more button', () => {
    const loadMore = jest.fn();
    const commentsAsReturned = {
      loadMore,
      hasMore: true,
      commentsList: makeComments([{ ideaId: 'idea1' }]).data,
    };
    const Wrapper = shallow(
      <UserComments comments={commentsAsReturned} theme={theme} />
    );
    expect(Wrapper).toMatchSnapshot();
  });
  it('loads more comments when the load more button is clicked', () => {
    const loadMore = jest.fn();
    const commentsAsReturned = {
      loadMore,
      hasMore: true,
      commentsList: makeComments([{ ideaId: 'idea1' }]).data,
    };
    const Wrapper = shallow(
      <UserComments comments={commentsAsReturned} theme={theme} />
    );
    Wrapper.find('UserComments__LoadMoreButton').simulate('click');
    expect(loadMore).toHaveBeenCalledTimes(1);
  });
});
