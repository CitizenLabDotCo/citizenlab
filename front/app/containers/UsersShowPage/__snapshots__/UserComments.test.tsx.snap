// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserComments /> renders correctly when empty 1`] = `
<UserComments__MessageContainer>
  <FormattedMessage
    defaultMessage="This user has not posted any comment yet."
    id="app.containers.UsersShowPage.noCommentsForUser"
  />
</UserComments__MessageContainer>
`;

exports[`<UserComments /> renders correctly when empty and own profile 1`] = `
<UserComments__MessageContainer>
  <FormattedMessage
    defaultMessage="You have not posted any comment yet."
    id="app.containers.UsersShowPage.noCommentsForYou"
  />
</UserComments__MessageContainer>
`;

exports[`<UserComments /> renders correctly when error 1`] = `
<UserComments__MessageContainer>
  <FormattedMessage
    defaultMessage="An error has occured, please try again later."
    id="app.containers.UsersShowPage.tryAgain"
  />
</UserComments__MessageContainer>
`;

exports[`<UserComments /> renders correctly with actual comments 1`] = `
<UserComments__Container
  className="e2e-profile-comments"
>
  <a11y__ScreenReaderOnly>
    <FormattedMessage
      defaultMessage="All the comments posted by this user"
      id="app.containers.UsersShowPage.invisibleTitleUserComments"
      tagName="h2"
    />
  </a11y__ScreenReaderOnly>
  <WrappedPostComments
    comments={
      [
        {
          "attributes": {
            "body_multiloc": {
              "en": "Just a comment",
            },
            "children_count": 0,
            "created_at": "2019-03-12T00: 00: 00.000Z",
            "downvotes_count": 0,
            "publication_status": "published",
            "updated_at": "2019-03-26T14: 32: 32.000Z",
            "upvotes_count": 0,
          },
          "id": "commentId",
          "relationships": {
            "author": {
              "data": {
                "id": "authorId",
                "type": "user",
              },
            },
            "parent": {
              "data": null,
            },
            "post": {
              "data": {
                "id": "idea1",
                "type": "idea",
              },
            },
            "user_vote": {
              "data": null,
            },
          },
          "type": "comment",
        },
      ]
    }
    key="idea1"
    postId="idea1"
    postType="idea"
  />
  <WrappedPostComments
    comments={
      [
        {
          "attributes": {
            "body_multiloc": {
              "en": "Just a comment",
            },
            "children_count": 0,
            "created_at": "2019-03-12T00: 00: 00.000Z",
            "downvotes_count": 0,
            "publication_status": "published",
            "updated_at": "2019-03-26T14: 32: 32.000Z",
            "upvotes_count": 0,
          },
          "id": "commentId",
          "relationships": {
            "author": {
              "data": {
                "id": "authorId",
                "type": "user",
              },
            },
            "parent": {
              "data": null,
            },
            "post": {
              "data": {
                "id": "idea2",
                "type": "idea",
              },
            },
            "user_vote": {
              "data": null,
            },
          },
          "type": "comment",
        },
        {
          "attributes": {
            "body_multiloc": {
              "en": "Just a comment",
            },
            "children_count": 0,
            "created_at": "2019-03-12T00: 00: 00.000Z",
            "downvotes_count": 0,
            "publication_status": "published",
            "updated_at": "2019-03-26T14: 32: 32.000Z",
            "upvotes_count": 0,
          },
          "id": "commentId",
          "relationships": {
            "author": {
              "data": {
                "id": "authorId",
                "type": "user",
              },
            },
            "parent": {
              "data": null,
            },
            "post": {
              "data": {
                "id": "idea2",
                "type": "idea",
              },
            },
            "user_vote": {
              "data": null,
            },
          },
          "type": "comment",
        },
      ]
    }
    key="idea2"
    postId="idea2"
    postType="idea"
  />
</UserComments__Container>
`;

exports[`<UserComments /> renders correctly with actual comments and load more button 1`] = `
<UserComments__Container
  className="e2e-profile-comments"
>
  <a11y__ScreenReaderOnly>
    <FormattedMessage
      defaultMessage="All the comments posted by this user"
      id="app.containers.UsersShowPage.invisibleTitleUserComments"
      tagName="h2"
    />
  </a11y__ScreenReaderOnly>
  <WrappedPostComments
    comments={
      [
        {
          "attributes": {
            "body_multiloc": {
              "en": "Just a comment",
            },
            "children_count": 0,
            "created_at": "2019-03-12T00: 00: 00.000Z",
            "downvotes_count": 0,
            "publication_status": "published",
            "updated_at": "2019-03-26T14: 32: 32.000Z",
            "upvotes_count": 0,
          },
          "id": "commentId",
          "relationships": {
            "author": {
              "data": {
                "id": "authorId",
                "type": "user",
              },
            },
            "parent": {
              "data": null,
            },
            "post": {
              "data": {
                "id": "idea1",
                "type": "idea",
              },
            },
            "user_vote": {
              "data": null,
            },
          },
          "type": "comment",
        },
      ]
    }
    key="idea1"
    postId="idea1"
    postType="idea"
  />
  <UserComments__Footer>
    <UserComments__LoadMoreButton
      height="50px"
      icon="refresh"
      onClick={[MockFunction]}
    >
      <FormattedMessage
        defaultMessage="Load more comments"
        id="app.containers.UsersShowPage.loadMoreComments"
      />
    </UserComments__LoadMoreButton>
  </UserComments__Footer>
</UserComments__Container>
`;
